### PR TITLE
Fix regression in automatic_face_movement_max_rotation_per_sec

### LINKED ISSUE #8451: Regression in automatic_face_movement_max_rotation_per_sec (and patch by p_gimeno)
Opening on behalf of p_gimeno.

"Values <= 0 should make the yaw change instant. This worked in 0.4.16 but was broken in 089f5945.
Per bug report by oil_boi_minetest on IRC."

Get patch here https://gitlab.com/pgimeno/minetest/merge_requests/5.patch
If desired i can make a PR for this.
```
diff --git a/src/client/content_cao.cpp b/src/client/content_cao.cpp
index 8643b5824b031f3b27e9e45485cab0d2fc3a3b23..22f62e6c3cf941c5adbf0af0b68f7cfbbe224dbb 100644
--- a/src/client/content_cao.cpp
+++ b/src/client/content_cao.cpp
@@ -1000,10 +1000,16 @@ void GenericCAO::step(float dtime, ClientEnvironment *env)
 
 		float target_yaw = atan2(m_velocity.Z, m_velocity.X) * 180 / M_PI
 				+ m_prop.automatic_face_movement_dir_offset;
-		float max_rotation_delta =
-				dtime * m_prop.automatic_face_movement_max_rotation_per_sec;
+		float max_rotation_per_sec =
+				m_prop.automatic_face_movement_max_rotation_per_sec;
+		if (max_rotation_per_sec > 0) {
+			float max_rotation_delta = dtime * max_rotation_per_sec;
+
+			wrappedApproachShortest(m_rotation.Y, target_yaw, max_rotation_delta, 360.f);
+		} else
+			// Negative values of ...max_rotation_per_sec mean disabled.
+			m_rotation.Y = target_yaw;
 
-		wrappedApproachShortest(m_rotation.Y, target_yaw, max_rotation_delta, 360.f);
 		rot_translator.val_current = m_rotation;
 
 		updateNodePos();
diff --git a/src/content_sao.cpp b/src/content_sao.cpp
index cb0a969ebb61d9640217e598198f51e35e48d456..0b19821d320fe9d85ad94767a28b569f61caddc5 100644
--- a/src/content_sao.cpp
+++ b/src/content_sao.cpp
@@ -457,11 +457,16 @@ void LuaEntitySAO::step(float dtime, bool send_recommended)
 
 			float target_yaw = atan2(m_velocity.Z, m_velocity.X) * 180 / M_PI
 				+ m_prop.automatic_face_movement_dir_offset;
-			float max_rotation_delta =
-					dtime * m_prop.automatic_face_movement_max_rotation_per_sec;
 
-			m_rotation.Y = wrapDegrees_0_360(m_rotation.Y);
-			wrappedApproachShortest(m_rotation.Y, target_yaw, max_rotation_delta, 360.f);
+			float max_rotation_per_sec =
+					m_prop.automatic_face_movement_max_rotation_per_sec;
+			if (max_rotation_per_sec > 0) {
+				float max_rotation_delta = dtime * max_rotation_per_sec;
+				m_rotation.Y = wrapDegrees_0_360(m_rotation.Y);
+				wrappedApproachShortest(m_rotation.Y, target_yaw, max_rotation_delta, 360.f);
+			} else
+				// Negative values of ...max_rotation_per_sec mean disabled.
+				m_rotation.Y = target_yaw;
 		}
 	}
 ```

### PR DESCRIPTION
**Pull request by paramat**
Values <= 0 should make the yaw change instant. This worked in 0.4.16 but was broken in 089f59458286.

Per bug report by oil_boi_minetest on IRC.
////////////////////

Closes #8451 
1st commit by p_gimeno.

### CODE REVIEW

File: src/client/content_cao.cpp [hunk 1]
```diff
@@ -1000,10 +1000,16 @@ void GenericCAO::step(float dtime, ClientEnvironment *env)
...
 				+ m_prop.automatic_face_movement_dir_offset;
-		float max_rotation_delta =
-				dtime * m_prop.automatic_face_movement_max_rotation_per_sec;
+		float max_rotation_per_sec =
+				m_prop.automatic_face_movement_max_rotation_per_sec;
+		if (max_rotation_per_sec > 0) {
+			float max_rotation_delta = dtime * max_rotation_per_sec;
+
+			wrappedApproachShortest(m_rotation.Y, target_yaw, max_rotation_delta, 360.f);
+		} else
```
Comment (SmallJoker) [inline src/client/content_cao.cpp]: Missing braces.

File: src/client/content_cao.cpp [hunk 2]
```diff
@@ -1000,10 +1000,16 @@ void GenericCAO::step(float dtime, ClientEnvironment *env)
...
 		float target_yaw = atan2(m_velocity.Z, m_velocity.X) * 180 / M_PI
 				+ m_prop.automatic_face_movement_dir_offset;
-		float max_rotation_delta =
-				dtime * m_prop.automatic_face_movement_max_rotation_per_sec;
+		float max_rotation_per_sec =
+				m_prop.automatic_face_movement_max_rotation_per_sec;
+		if (max_rotation_per_sec > 0) {
+			float max_rotation_delta = dtime * max_rotation_per_sec;
+
+			wrappedApproachShortest(m_rotation.Y, target_yaw, max_rotation_delta, 360.f);
```
Comment (SmallJoker) [inline src/client/content_cao.cpp]: Please inline `max_rotation_delta`

File: src/content_sao.cpp [hunk 3]
```diff
@@ -457,11 +457,16 @@ void LuaEntitySAO::step(float dtime, bool send_recommended)
...
 
-			m_rotation.Y = wrapDegrees_0_360(m_rotation.Y);
-			wrappedApproachShortest(m_rotation.Y, target_yaw, max_rotation_delta, 360.f);
+			float max_rotation_per_sec =
+					m_prop.automatic_face_movement_max_rotation_per_sec;
+			if (max_rotation_per_sec > 0) {
+				float max_rotation_delta = dtime * max_rotation_per_sec;
+				m_rotation.Y = wrapDegrees_0_360(m_rotation.Y);
+				wrappedApproachShortest(m_rotation.Y, target_yaw, max_rotation_delta, 360.f);
+			} else
```
Comment (SmallJoker) [inline src/content_sao.cpp]: Same as in content_cao

**Comment by paramat** [comment]
IRC http://irc.minetest.net/minetest-dev/2019-04-07#i_5526113

17:02 p_gimeno   Krock: About this: https://github.com/minetest/minetest/issues/8451#issuecomment-480522773 I moved the calculation of max_rotation_delta inside the 'if' so that it doesn't need to be evaluated in the other 'if' branch where it's not needed
17:03 Krock      p_gimeno: oh right, I see
17:04 Krock      the impact on performance is very tiny so having the multiplication outside wouldn't hurt either. However, I think the inner variable should be inlined since it's only used once
17:05 p_gimeno   fine with me
17:05 p_gimeno   also the new variable is just a shortened version of the other setting, which could be used in its place

File: src/client/content_cao.cpp (hunk 1 above)
Comment (paramat) [reply src/client/content_cao.cpp]: Done.

File: src/client/content_cao.cpp (hunk 2 above)
Comment (paramat) [reply src/client/content_cao.cpp]: Done.

File: src/content_sao.cpp (hunk 3 above)
Comment (paramat) [reply src/content_sao.cpp]: Done.

**Comment by paramat** [comment]
> Docs also need adjusting to No limit if value <= 0.

Done and requests attended to.
Compiles but i haven't tested.

**Comment by paramat** [comment]
Sorry p+gimeno i just saw you have updated your patch, however it does have a mistake:
```
  |   | +		if (max_rotation_per_sec > 0)
  |   | +			wrappedApproachShortest(m_rotation.Y, target_yaw,
  |   | +					dtime * max_rotation_per_sec, 360.f);
  |   | +		else {
```
Braces error.

**Comment by rubenwardy** [comment]
My approval still holds

**Comment by paramat** [comment]
I saw on IRC there seems to be some confusion.
What i meant is, and what our code style rule is: 'don't mix use of braces and use of no braces in one conditional'.

**Comment by paramat** [comment]
IRC

18:43 rubenwardy "code style reviews are stupid, that's what the tools are for. Look for real issues first"
18:48 nerzhul    linter is here to prvent that if we just use it correctly :p
18:52 rubenwardy yeah

This is very silly =) Code style review is obviously very important.
The 'tools' we have, the linter, misses most codestyle issues and complains endlessly about non-problems. Our automatic codestyle checking is almost completely incompetant.

**Comment by rubenwardy** [comment]
Code style is much less important than working and correct code. Code style review is incredibly easy, and creates a false comfort of "oh the braces are all in the right place, code must be correct".

I'm all for a rule where you shouldn't suggest code style changes until you review the things that actually matter, unless it's actually unreadable (which it very rarely is)

**Comment by paramat** [comment]
The don't mix thing isn't in our codestyle guidelines, but "The coding style is based on the Linux kernel code style" and it's in there https://www.kernel.org/doc/html/latest/process/coding-style.html
> Code style is much less important than working and correct code.

No, not 'much less', maybe 'slightly less', readable code is extremely important and essential, obviously.
> and creates a false comfort of "oh the braces are all in the right place, code must be correct".

In who does it create this false comfort? Not in me, not in any core dev, only in a fool, so this is pointless to write.
> I'm all for a rule where you shouldn't suggest code style changes until you review the things that actually matter.

That would be silly and there's no benefit in doing this, and there's no harm in pointing out codestyle issues at any point. If they're not pointed out when noticed they are likely to be forgotten.

If a core dev doesn't want to get deep into fully understanding the code of a PR, they should be free to contribute and help in any way they can, by pointing out codestyle issues. Not allowing this is a waste and results in worse code.
> The linter should be fixed to work so we don't waste time in this trap

There is no trap.

**Comment by rubenwardy** [comment]
Code style is a small part of readable code. I have noticed that reviewers tend to go full into code style review without first thinking about the use of a PR, reading any questions, and looking for obvious issues. It is incredibly annoying when people make code reviews without even reading the opening post or any big issues that need feedback.

You have been subject to this trap, I have been subject to this trap. And remember, fool and core dev aren't mutually exclusive :)

**Comment by paramat** [comment]
Nerzhul does it too but it's ok.
> It is incredibly annoying when people make code reviews without even reading the opening post or any big issues that need feedback.

It doesn't bother me, i don't see why it should bother anyone, every little bit of review is helpful.
A core dev may not have the energy or knowledge to review the PR in depth (i am often in this situation), so may help out with good intentions in any way they can.

Neither does it happen much, reviewers very often deal with the PR in full.

I don't like all this fuss, nonsense and negativity over a small and helpful bit of review. Through doing this i am only helping the contributor understand MT codestyle, which will benefit their future PRs.

**Comment by ClobberXD** [comment]
> A core dev may not have the energy or knowledge to review the PR in depth (i am often in this situation), so may help out with good intentions in any way they can.

There are many-a-times where I tend to do code-style-only reviews as well, and I wholly agree with this. 

(This discussion is probably going offtopic...)

**Pull request merged**